File: python/Q211_design_add_and_search_words_data_structure.py
#
# @lc app=leetcode id=211 lang=python3
#
# [211] Design Add and Search Words Data Structure
#


# @lc code=start
class TrieNode:
    def __init__(self) -> None:
        self.children: dict[str, TrieNode] = dict()
        self.end: bool = False


class WordDictionary:
    def __init__(self) -> None:
        self.root: TrieNode = TrieNode()

    def addWord(self, word: str) -> None:
        current: TrieNode = self.root

        for c in word:
            if c not in current.children:
                current.children[c] = TrieNode()

            current = current.children[c]

        current.end = True

    def search_with_node(self, word: str, node: TrieNode, index: int) -> bool:
        if index != len(word):
            if word[index] == ".":
                return any(
                    self.search_with_node(word, child, index + 1)
                    for child in node.children.values()
                )

            if word[index] in node.children:
                return self.search_with_node(
                    word, node.children[word[index]], index + 1
                )
            else:
                return False
        else:
            return node.end

    def search(self, word: str) -> bool:
        return self.search_with_node(word, self.root, 0)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
# @lc code=end

File: rust/src/leetcode/Q211_design_add_and_search_words_data_structure.rs
/*
 * @lc app=leetcode id=211 lang=rust
 *
 * [211] Design Add and Search Words Data Structure
 */

// @lc code=start
use std::collections::HashMap;

struct TrieNode {
    children: HashMap<char, Box<TrieNode>>,
    word: bool,
}

impl TrieNode {
    #[inline]
    fn new() -> Self {
        TrieNode {
            children: HashMap::new(),
            word: false,
        }
    }
}

struct WordDictionary {
    start: TrieNode,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl WordDictionary {
    fn new() -> Self {
        WordDictionary {
            start: TrieNode::new(),
        }
    }

    fn add_word(&mut self, word: String) {
        let mut current: &mut TrieNode = &mut self.start;

        for c in word.chars() {
            current = current
                .children
                .entry(c)
                .or_insert(Box::new(TrieNode::new()));
        }

        current.word = true;
    }

    fn search_range(chars: &Vec<char>, node: &TrieNode, index: usize) -> bool {
        if index != chars.len() {
            if chars[index] == '.' {
                return node
                    .children
                    .values()
                    .any(|child| Self::search_range(chars, child, index + 1));
            }

            if let Some(next_node) = node.children.get(&chars[index]) {
                Self::search_range(chars, next_node, index + 1)
            } else {
                false
            }
        } else {
            node.word
        }
    }

    fn search(&self, word: String) -> bool {
        let chars: Vec<char> = word.chars().collect();
        Self::search_range(&chars, &self.start, 0)
    }
}

// /**
//  * Your WordDictionary object will be instantiated and called as such:
//  * let obj = WordDictionary::new();
//  * obj.add_word(word);
//  * let ret_2: bool = obj.search(word);
//  */
// @lc code=end
